contributor-docs/adrs/adr-005-box-sx.md

```diff
@@ -0,0 +1,83 @@
+# ADR 005: Use Box as a building block for all other components
+
+## Status
+
+Proposed
+
+## Context
+
+In Primer React and consuming applications, we use many different patterns for creating React components. Two common patterns are:
+
+1. Creating components with styled-components
+
+    ```tsx
+    const Avatar = styled.img.attrs<StyledAvatarProps>(props => ({
+      height: props.size,
+      width: props.size
+    }))<StyledAvatarProps>`
+      display: inline-block;
+      overflow: hidden;
+      line-height: ${get('lineHeights.condensedUltra')};
+      border-radius: ${props => getBorderRadius(props)};
+      ${sx}
+    `
+    ```
+
+    [Show full code example →](https://github.com/primer/react/pull/2019/files?diff=split&w=0)
+
+2.  Creating components with Box
+
+    ```tsx
+    const Avatar: React.FC<AvatarProps> = ({size = 20, alt = '', square = false, sx = {}, ...props}) => {
+      const styles = {
+        display: 'inline-block',
+        overflow: 'hidden',
+        lineHeight: 'condensedUltra',
+        borderRadius: getBorderRadius({size, square})
+      }
+
+      return (
+        <Box
+          as="img"
+          alt={alt}
+          sx={merge(styles, sx as SxProp)} // styles needs to merge with props.sx
+          {...props}
+        />
+      )
+    }
+    ```
+
+    [Show full code example →](https://github.com/primer/react/pull/2019/files?diff=split&w=0)
+
+&nbsp;
+
+## Decision
+
+Prefer using method #2: Creating components with Box for the following reasons:
+
+- Better authoring experience with typescript. With Box, we can improve the API and autocomplete for consuming primitives.
+- The styling library (i.e. styled-components) becomes an implementation detail and can be changed later with minimal breaking changes for consumers. (Avoids leaky abstractions)
```

Comment:
In what cases what `styled-comopnents` be appropriate?

Reply:
> In what cases what styled-comopnents be appropriate?

I want to say that if you're using primer primitives, then Box + sx is always recommended over styled-components. 

There is of course a slight syntax difference between string with styled-components and an object with Box, which can be a tradeoff. The planned improvements with Box should make that tradeoff easy to make though.

@colebemis any use cases where this wouldn't be true?

Reply:
Could the additional [runtime overheads ](https://github.com/github/primer/discussions/755#discussioncomment-2452177) to using Box + sx be highlighted in this ADR as a known risk? Particularly as we're recommending authors use the costlier implementation approach at this time.

Reply:
@siddharthkp @colebemis related to 

> I don't think there's anything you can do with styled-components that you can't do with Box + sx.

Should Primer expose a replacement for the styled-components built in version of `keyframes` and `createGlobalStyle` (even if they're just re-exports for now)?  I believe they may be the only apis from styled-components that consumers would still need to call directly.

This might enable applications that are "all in" on using primer primitives for styling, and allow for things like an eslint plugin that bans internal usage of styled-components directly, directing consumers to the primer-based alternatives?

Reply:
I like that idea 🤔 

Just to make sure, could this result in multiple versions of styled-components on the page because of version mismatch?

Reply:
As long as it's still just a peer of primer I don't think there's any risk of multiple versions for sc.

I think that the only limit on styled components be via an eslint rule saying, use box and ..... Instead of using styled components directly, which should provide a nice way to nudge other sc usage towards primer apis